Source/bmalloc/libpas/src/libpas/pas_root.c

```diff
@@ -389,5 +390,101 @@ pas_root* pas_root_ensure_for_libmalloc_enumeration(void)
     return root;
 }
 
+#define PAS_SYSTEM_COMPACT_POINTER_SIZE 4
+
+#if PAS_CPU(ADDRESS64)
+#if PAS_ARM64 && PAS_OS(DARWIN)
+#if MACH_VM_MAX_ADDRESS_RAW < (1ULL << 36)
+#define PAS_HAVE_36BIT_ADDRESS 1
+#endif
+#endif
+#endif // PAS_CPU(ADDRESS64)
+
+static inline vm_address_t decode_system_compact_pointer(pas_root* root, uint32_t value)
+{
+    PAS_UNUSED_PARAM(root);
+#if PAS_HAVE(36BIT_ADDRESS)
+    /* Our address region is 36-bit, thus we encode a 16-byte aligned pointer into 32-bit value by shifting it by 4. */
+    return (vm_address_t)(((uintptr_t)value) << 4);
+#else
+    /* The other platforms are not supporting CompactPtr. */
+    PAS_UNUSED_PARAM(value);
```

Comment:
It's a pity (and wasteful) to make `pas_root_visit_conservative_candidate_pointers_in_address_range` call `decode_system_compact_pointer` repeatedly when NOT `PAS_HAVE(36BIT_ADDRESS)`, only to never find any pointers.  Can you have `pas_root_visit_conservative_candidate_pointers_in_address_range` just return early if NOT `PAS_HAVE(36BIT_ADDRESS)` or define and check a `has_pointer_encoding_scheme_to_decode` boolean?  With the boolean, we can make it dependent on more than one encoding scheme in the future if we ever add other encoding schemes.

Reply:
Done.